### PR TITLE
Fix offset legend tightbbox

### LINKED ISSUE #10194: legend is not present in the generated image if I use 'tight' for bbox_inches
## summary of the problem
I tried to put legend outside of the axes. If I use `fig.legend()` method to produce legend and use `bbox_inches='tight'` in `savefig()` method. The legend is not present in the produced image (using the argument `bbox_extra_artists` also does not work). If I use `ax.legend()` instead, the legend is present in the saved image.

**Code for reproduction**
```python
import matplotlib.pyplot as plt
import numpy as np
import matplotlib as mpl

x = np.arange(-5, 5, 0.1)
y1 = np.sin(x)
y2 = np.cos(x)

y3 = np.sin(2*x)
y4 = np.cos(2*x)

fig, (ax1, ax2) = plt.subplots(ncols=1, nrows=2, figsize=(10, 6))

ax1.plot(x, y1, label='sin')
ax1.plot(x, y2, label='cos')

ax2.plot(x, y3, label='sin')
ax2.plot(x, y4, label='cos')

handles, labels = ax1.get_legend_handles_labels()
# following code does not work
legend = fig.legend(handles, labels, loc='lower left', ncol=2, frameon=False,
              bbox_to_anchor=(0.12, 0.88))
plt.savefig('test.jpg', bbox_extra_artists=[legend,], bbox_inches='tight')

# following code works
# legend = ax1.legend(handles, labels, ncol=2, frameon=False,
#                     loc='lower left', bbox_to_anchor=(-0.01, 1.2))
# plt.savefig('test.jpg', bbox_inches='tight')
plt.show()
```
## outcome
The saved images are different if I use different legend method.
### using `fig.legend` method
image

### using `ax.legend` method
image

**Matplotlib version**
  * Operating system: Windows 8.1 and Linux (CentOS 7) (This problem is maybe platform independent)
  * Matplotlib version:  2.0.2(Linux), 2.0.0 (Windows)
  * Matplotlib backend: Qt5Agg
  * Python version: 3.5.2 (Windows), 3.6.2 (Linux)

The matplotlib and python are all distributed with Anaconda.

## quesiton
Is this a bug or something I do not know. Maybe there is difference between `fig.legend` and `ax.legend` method?

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes #10194

legend, if it receives a "None" for `bbox_transform` positions the legend relative to its `parent.bbox`, which usually works fine, but if you mess with the `transFigure`, as we do when `bbox_inches='tight'` is called, then the new transFigure is not used (everything is still relative to the bbox of the figure, which doesn't change).  

This tiny change just sets a default transform for figure legends to be the figure's transform.  

and adds a test...

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
This fixes  pretty old bug marked as 3.3, so marking 3.3.  But feel free to push forward to 3.4

File: lib/matplotlib/tight_bbox.py [hunk 1]
```diff
@@ -45,6 +50,9 @@ def restore_bbox():
         fig.set_tight_layout(orig_tight_layout)
         fig.transFigure._boxout = _boxout
         fig.transFigure.invalidate()
+        for l in fig.legends:  # re-set the anchor for legends...
+            anch = legend_anchors[l].transformed(fig.transFigure.inverted())
+            l.set_bbox_to_anchor(anch)
```
Comment (QuLogic) [inline lib/matplotlib/tight_bbox.py:55]: This kind of assumes what the original `transform` argument was, doesn't it?

Also, the setter will set the artist `stale`. Maybe not using the setter is better?

Comment (jklymak) [reply lib/matplotlib/tight_bbox.py:55]: Yep, it assumes no transform.  Does the transform argument work at all?

Comment (QuLogic) [reply lib/matplotlib/tight_bbox.py:55]: There's one example in the constrained layout guide:
https://github.com/matplotlib/matplotlib/blob/9ccdb1ddcd6250fac950ddcab97c41f59f68dd72/tutorials/intermediate/constrainedlayout_guide.py#L246-L247

Comment (jklymak) [reply lib/matplotlib/tight_bbox.py:55]: Ohhhh, I see.   OK this is a much easier and more robust fix.

**Comment by QuLogic** [comment]
It looks like a different test might have broken?

**Comment by jklymak** [comment]
Lots broke!  I guess thats helpful, because I'm curious why...

**Comment by jklymak** [comment]
breakage was trivial:  There was a test for the arguments legend was being called with, and since we are explicitly passing `bbox_transform` we need to include it in that list.

**Pull request merged**